crates/hyperqueue/src/stream/reader/streamdir.rs

```diff
@@ -0,0 +1,464 @@
+use crate::client::commands::reader::{CatOpts, Channel, ExportOpts, ShowOpts};
+use crate::common::arraydef::IntArray;
+use crate::common::error::HqError;
+use crate::server::event::bincode_config;
+use crate::transfer::stream::{ChannelId, StreamChunkHeader};
+use crate::worker::streamer::{StreamFileHeader, STREAM_FILE_HEADER};
+use crate::{JobId, JobTaskId, Set};
+use bincode::Options;
+use chrono::{DateTime, Utc};
+use colored::{Color, Colorize};
+use lru::LruCache;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use smallvec::SmallVec;
+use std::collections::BTreeMap;
+use std::fs::File;
+use std::io::{BufReader, BufWriter, Read, Seek, Write};
+use std::num::NonZeroUsize;
+use std::ops::Deref;
+use std::path::{Path, PathBuf};
+use tako::InstanceId;
+
+#[derive(Clone)]
+pub struct ChunkInfo {
+    time: DateTime<Utc>,
+    position: u64,
+    size: u32, // Currently chunk is actually limited to 128kB
+}
+
+pub struct InstanceInfo {
+    instance_id: InstanceId,
+    channels: [Vec<ChunkInfo>; 2],
+    file_idx: usize,
+    finished: bool,
+}
+
+#[derive(Default)]
+pub struct TaskInfo {
+    instances: SmallVec<[InstanceInfo; 1]>,
+}
+
+impl InstanceInfo {
+    fn channel_size(&self, channel_id: ChannelId) -> u64 {
+        self.channels[channel_id as usize]
+            .iter()
+            .map(|x| x.size as u64)
+            .sum()
+    }
+}
+
+impl TaskInfo {
+    pub fn last_instance(&self) -> &InstanceInfo {
+        self.instances.last().unwrap()
+    }
+
+    pub fn superseded(&self) -> impl Iterator<Item = &InstanceInfo> {
+        self.instances[0..self.instances.len() - 1].iter()
+    }
+
+    pub fn instance_mut(&mut self, instance_id: InstanceId) -> Option<&mut InstanceInfo> {
+        self.instances
+            .iter_mut()
+            .find(|info| info.instance_id == instance_id)
+    }
+
+    pub fn instance(&self, instance_id: InstanceId) -> Option<&InstanceInfo> {
+        self.instances
+            .iter()
+            .find(|info| info.instance_id == instance_id)
+    }
+}
+
+pub struct StreamDir {
```

Comment:
Could you please add some comments either to this struct or to the whole file, to provide some high-level explanation of what is this good for and (mainly) how does it work?